### ISSUE TITLE
[Tier 1-Task 3] Implement ONNX Model Export Pipeline with Cross-Platform Validation

### ISSUE THREAD

**Issue by harris-boyce** [harris-boyce/movie-recs#1]
# 🤖 ONNX Model Export & Cross-Platform Validation

**Epic:** MovieRecs Brownfield Integration - Tier 1  
**Task Priority:** 3/9 (Quick Win)  
**Estimated Effort:** 1-3 days  
**Created:** 2025-09-03

### Primary Goals
- [ ] Establish reliable ML model → ONNX export pipeline
- [ ] Validate cross-platform model execution (Python → .NET)
- [ ] Create baseline model that proves the entire ML→API integration
- [ ] Enable confident development of ASP.NET Core model serving

### Success Criteria
- [ ] Dummy/baseline model exports to valid ONNX format
- [ ] ONNX model loads and runs in both Python and .NET environments
- [ ] Inference results match between platforms (within tolerance)
- [ ] Model metadata includes version, schema, and feature definitions
- [ ] Export pipeline is fully automated and reproducible
- [ ] Performance benchmarks established for different model sizes

### Baseline Model Specifications
- [ ] **Model Type:** Start with Linear Regression (multi-output for multiple taste features)
- [ ] **Input Features:** Basic movie features from data pipeline
  - Genre one-hot encoding (20 dimensions)
  - Runtime (normalized)
  - Release decade (one-hot, 10 dimensions)
  - Rating average/count (normalized)
- [ ] **Output:** Movie "taste scores" (3-5 dimensional taste profile)
  - quality_score: 0-1 (predicted rating quality)
  - genre_preference: 0-1 (genre appeal)
  - complexity_score: 0-1 (narrative complexity)
  - mainstream_appeal: 0-1 (broad vs niche)

### Model Training Pipeline
```python
# Expected model interface
class MovieTasteModel:
    def __init__(self, input_dim: int = 35, output_dim: int = 4):
        pass
    
    def train(self, X: np.ndarray, y: np.ndarray) -> TrainingResults:
        pass
    
    def predict(self, X: np.ndarray) -> np.ndarray:
        pass
    
    def export_onnx(self, path: str, metadata: dict) -> None:
        pass
```

### ONNX Export Requirements
- [ ] **Format Compliance:** ONNX 1.12+ compatible
- [ ] **Input Specification:**
  - Name: "movie_features"
  - Shape: [batch_size, 35] (dynamic batch dimension)
  - Type: float32
- [ ] **Output Specification:**
  - Name: "taste_scores"  
  - Shape: [batch_size, 4]
  - Type: float32
- [ ] **Metadata Inclusion:**
  - Model version and training date
  - Feature names and expected ranges
  - Output interpretation guide
  - Performance metrics (accuracy, training time)

### Cross-Platform Validation
- [ ] **Python Validation:**
  - Load ONNX with onnxruntime-python
  - Run inference on test dataset
  - Compare with original PyTorch/sklearn predictions
- [ ] **C#/.NET Validation:**
  - Load ONNX with Microsoft.ML.OnnxRuntime
  - Run identical test inputs
  - Verify outputs match Python (tolerance: 1e-6)
- [ ] **Performance Benchmarking:**
  - Latency measurements (single prediction + batch)
  - Memory usage profiling
  - Throughput testing

### Directory Structure
```
python-ml/
├── src/
│   ├── models/
│   │   ├── base_model.py        # Abstract model interface
│   │   ├── linear_taste_model.py # Initial linear regression
│   │   └── model_registry.py    # Model versioning utilities
│   ├── training/
│   │   ├── train.py            # Training orchestration
│   │   ├── evaluation.py       # Model evaluation metrics
│   │   └── hyperparameters.py  # Config management
│   ├── export/
│   │   ├── onnx_exporter.py    # ONNX conversion utilities
│   │   ├── validation.py       # Cross-platform testing
│   │   └── metadata.py         # Model metadata management
│   └── utils/
│       ├── feature_engineering.py
│       └── data_loaders.py
├── tests/
│   ├── test_training.py        # Training pipeline tests
│   ├── test_onnx_export.py     # Export functionality tests
│   ├── test_cross_platform.py  # Python vs .NET validation
│   └── fixtures/
│       ├── sample_features.json
│       └── expected_outputs.json
├── models/                     # Output directory
│   ├── movie_taste.onnx
│   ├── model_metadata.json
│   └── performance_benchmarks.json
├── configs/
│   ├── model_config.yaml       # Training configuration
│   └── export_config.yaml      # Export settings
└── scripts/
    ├── train_and_export.py     # End-to-end script
    └── validate_onnx.py        # Cross-platform validation
```

### Configuration Management
```yaml
# model_config.yaml
model:
  type: "linear_regression"
  input_features: 35
  output_features: 4
  
training:
  test_size: 0.2
  random_state: 42
  validation_strategy: "stratified"
  
export:
  onnx_version: "1.12"
  opset_version: 14
  optimize: true
  include_metadata: true
  
validation:
  tolerance: 1e-6
  test_batch_sizes: [1, 10, 100]
  benchmark_iterations: 1000
```

### Model Metadata Schema
```json
{
  "model_info": {
    "name": "movie_taste_baseline_v1",
    "version": "1.0.0",
    "created_date": "2025-01-XX",
    "model_type": "linear_regression",
    "framework": "scikit-learn"
  },
  "input_schema": {
    "features": [
      {"name": "genre_action", "index": 0, "type": "float32", "range": [0, 1]},
      {"name": "genre_drama", "index": 1, "type": "float32", "range": [0, 1]},
      {"name": "runtime_normalized", "index": 2, "type": "float32", "range": [-2, 2]}
    ]
  },
  "output_schema": {
    "predictions": [
      {"name": "quality_score", "index": 0, "interpretation": "Predicted quality rating 0-1"},
      {"name": "genre_preference", "index": 1, "interpretation": "Genre appeal score 0-1"}
    ]
  },
  "performance": {
    "training_accuracy": 0.78,
    "validation_accuracy": 0.75,
    "training_time_seconds": 12.3,
    "model_size_mb": 0.1
  }
}
```

### Unit Tests
- [ ] Model training with synthetic data
- [ ] ONNX export format validation
- [ ] Metadata generation and parsing
- [ ] Feature preprocessing consistency

### Integration Tests  
- [ ] End-to-end training → export → validation pipeline
- [ ] Cross-platform prediction comparison
- [ ] Performance benchmark validation
- [ ] CI environment compatibility

### Cross-Platform Tests
- [ ] Python ONNX loading and inference
- [ ] C# ONNX loading and inference  
- [ ] Numerical accuracy comparison
- [ ] Performance parity validation

### Stress Tests
- [ ] Large batch prediction (1000+ samples)
- [ ] Memory usage under load
- [ ] Concurrent model loading
- [ ] Invalid input handling

### Code Artifacts
- [ ] Complete training and export pipeline
- [ ] Cross-platform validation scripts
- [ ] Comprehensive test suite (>95% coverage)
- [ ] CI/CD pipeline integration

### Model Artifacts  
- [ ] `movie_taste.onnx` - Trained ONNX model
- [ ] `model_metadata.json` - Complete model specification
- [ ] `performance_benchmarks.json` - Performance metrics
- [ ] Training logs and evaluation reports

### Validation Artifacts
- [ ] Cross-platform test results
- [ ] Performance comparison reports
- [ ] Example inference code (Python + C#)
- [ ] Integration testing documentation

### Technical Risks
- **Risk:** ONNX export fails due to unsupported operations
- **Mitigation:** Start with simple linear model + validate export early
- **Risk:** Cross-platform numerical differences
- **Mitigation:** Use consistent data types + tolerance-based comparison
- **Risk:** Performance degradation in ONNX vs native
- **Mitigation:** Benchmark early + optimize export settings

### Integration Risks  
- **Risk:** .NET runtime issues with ONNX model
- **Mitigation:** Test with exact ASP.NET Core target framework
- **Risk:** Model too large for deployment constraints
- **Mitigation:** Size monitoring + compression options

## ✅ **Definition of Done**

- [ ] Baseline model trains successfully on processed data
- [ ] ONNX export produces valid model file
- [ ] Cross-platform validation passes (Python ↔ .NET)
- [ ] Performance benchmarks within acceptable ranges
- [ ] Model metadata complete and validated
- [ ] All tests pass in CI environment
- [ ] Integration with downstream ASP.NET Core API verified
- [ ] Documentation complete with usage examples

### Prerequisites
- [ ] Task 2: Data Pipeline completed (need processed training data)
- [ ] Python ML environment with scikit-learn/PyTorch
- [ ] ONNX runtime packages (Python + .NET)

### Dependent Tasks
- This task enables:
  - ASP.NET Core Model API development
  - Model serving infrastructure
  - Frontend integration testing

---

## 🤖 **Implementation Prompts Section**
*To be filled in after specification approval*

### Claude Code Prompts
```markdown
[RESERVED SPACE]

After specification approval, add detailed Claude Code prompts here for:
1. Baseline model implementation (linear regression)
2. ONNX export pipeline with metadata
3. Cross-platform validation framework  
4. Performance benchmarking suite
5. CI/CD integration scripts
```

## 📝 **Implementation Notes**
*Add implementation decisions, alternative approaches considered, and lessons learned*

---
**Issue Template Version:** 1.0  
**Next Review:** [Date after spec approval]  
**Assignee:** [TBD]  
**Epic Link:** MovieRecs-Tier1
